src/main/java/com/dotcms/tomcat/redissessions/RedisSessionManager.java

```diff
@@ -479,63 +541,99 @@ protected boolean saveInternal(final Jedis jedis, final Session session, final b
+            this.setSessionExpiration(sessionId);
         } catch (final IOException e) {
-            log.error("An error occurred when saving Session " + session + ": " + e.getMessage(), e);
+            log.error("An error occurred when saving Session '" + sessionId + "': " + e.getMessage());
+            log.debug(e);
             throw e;
-        } finally {
-            return error;
         }
     }
 
+    /**
+     * Sets the expiration time for the newly created Session, as this process will be completely handled by Redis. It
+     * uses the exact same session timeout value configured in Tomcat.
+     * <p>If the value for the {@code DOT_DOTCMS_CLUSTER_ID} is specified, it'll be used to prefix they key. Doing this
+     * will allow multiple clusters to share the same Session Redis Store.</p>
+     *
+     * @param sessionId The ID of the Session whose TTL is being set.
+     */
+    private void setSessionExpiration(final String sessionId) {
+        final String prefixedKey = this.prefix + sessionId;
+        this.jedisPool.expire(prefixedKey.getBytes(), this.getTomcatSessionTimeoutInSeconds());
+    }
+
     @Override
-    public void remove(Session session) {
-        remove(session, false);
+    public void remove(final Session session) {
+        this.remove(session, false);
     }
 
     @Override
-    public void remove(Session session, boolean update) {
-        log.debug("Removing session ID: " + session.getId());
-        try (final Jedis jedis = acquireConnection()) {
-            jedis.del(session.getId());
+    public void remove(final Session session, final boolean update) {
+        if (this.isSessionPersistable(session)) {
+            log.debug("Removing session ID: " + session.getId());
+            this.del(session.getId());
         }
+        super.remove(session, update);
     }
 
+    /**
+     * This method is called by the {@link RedisSessionHandlerValve} after request has been processed. It takes care of
+     * optionally saving the current Session -- if it's still valid -- or removing it in case it is not.
+     */
     public void afterRequest() {
         final RedisSession redisSession = currentSession.get();
-        if (redisSession != null) {
-            try {
+        if (!this.isSessionPersistable(redisSession)) {
+            if (null != redisSession) {
                 if (redisSession.isValid()) {
-                    log.debug("Request with session completed, saving session: " + redisSession.getId());
-                    this.save(redisSession, getAlwaysSaveAfterRequest());
+                    super.add(redisSession);
                 } else {
-                    log.debug("HTTP Session has been invalidated, removing: " + redisSession.getId());
-                    this.remove(redisSession);
+                    super.remove(redisSession);
+                    currentSession.remove();
```

Comment:
All the threadlocal removals need to be in a finally block

Reply:
Done!